bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java

```diff
@@ -0,0 +1,299 @@
+        } catch (IOException | InterruptedException e) {
+            throw new AndroidDebugBridgeDeviceException("Can not open adb connection " + ip + ":" + port);
+        }
+        connection = adbConnection;
+        socket = sock;
+    }
+
+    private String runAdbShell(String... args)
+            throws IOException, InterruptedException, AndroidDebugBridgeDeviceException {
+        var adb = connection;
+        if (adb == null) {
+            throw new AndroidDebugBridgeDeviceException("Device not connected");
+        }
+        synchronized (adb) {
+            var byteArrayOutputStream = new ByteArrayOutputStream();
+            var cmd = String.join(" ", args);
+            logger.debug("{} - shell:{}", ip, cmd);
+            try {
+                var stream = adb.open("shell:" + cmd);
+                do {
+                    byteArrayOutputStream.writeBytes(stream.read());
+                } while (!stream.isClosed());
+            } catch (IOException e) {
+                var message = e.getMessage();
+                if (message != null && !message.equals("Stream closed"))
+                    throw e;
+            }
+            return byteArrayOutputStream.toString(StandardCharsets.US_ASCII);
+        }
+    }
+
+    private static AdbBase64 getBase64Impl() {
+        return bytes -> new String(Base64.getEncoder().encode(bytes));
+    }
+
+    private static AdbCrypto loadKeyPair(String pubKeyFile, String privKeyFile)
+            throws NoSuchAlgorithmException, IOException, InvalidKeySpecException {
+        File pub = new File(pubKeyFile);
+        File priv = new File(privKeyFile);
+        AdbCrypto c = null;
+        // load key pair
+        if (pub.exists() && priv.exists()) {
+            try {
+                c = AdbCrypto.loadAdbKeyPair(getBase64Impl(), priv, pub);
+            } catch (IOException ignored) {
+                // Keys don't exits
+            }
+        }
+        if (c == null) {
+            // generate key pair
+            c = AdbCrypto.generateAdbKeyPair(getBase64Impl());
+            c.saveAdbKeyPair(priv, pub);
+        }
+        return c;
+    }
+
+    public void disconnect() {
+        var adb = connection;
+        var sock = socket;
+        if (adb != null) {
+            try {
+                adb.close();
+            } catch (IOException ignored) {
+            }
+            connection = null;
+        }
+        if (sock != null) {
+            try {
+                sock.close();
+            } catch (IOException ignored) {
+            }
+            socket = null;
+        }
+    }
+
+    static class AndroidDebugBridgeDeviceException extends Exception {
+        public AndroidDebugBridgeDeviceException(String message) {
+            super(message);
+        }
+    }
+
+    static class AndroidDebugBridgeDeviceReadException extends Exception {
+        public AndroidDebugBridgeDeviceReadException(String message) {
+            super(message);
+        }
+    }
```

Comment:
Since these exceptions are used outside of this class, please move these exceptions to their own files.